content/recipes/interoperability/Storing text data in a knowledge graph.md

```diff
@@ -0,0 +1,306 @@
+    data = pd.DataFrame(requests.get(url).json().get('response').get('docs'))
+    # Return the first id (A simplistic non-perfect way for mapping)
+    return data["label"][0]
+  except:
+    return 'id-less'
+```
+After redifining the Rebel spaCy component, we include it into our pipeline:
+
+```python
+rel_ext = spacy.load('en_core_web_sm', disable=['ner', 'lemmatizer', 'attribute_rules', 'tagger'])
+rel_ext.add_pipe("rebel", config={
+    'device':DEVICE, # Number of the GPU, -1 if want to use CPU
+    'model_name':'Babelscape/rebel-large'} # Model used
+    )
+```
+
+Now we can text the pipeline on two simple sentances:
+```python
+input_text = "High fever is very dangerous. It can be treated with paracetamol."
+
+#coreference model
+coref_text = coref(input_text)._.resolved_text
+print(coref_text)
+#output: High fever is very dangerous. High fever can be treated with paracetamol.
+
+#entity and relations extraction
+doc = rel_ext(coref_text)
+
+for value, rel_dict in doc._.rel.items():
+    print(f"{value}: {rel_dict}")
+
+#output:
+#0440ea848947d2677bc11443f99f20f67ce0a1bc: {'relation': 'subclass of', 'head_span': {'text': 'High fever', 'id': 'High Grade Fever'}, 'tail_span': {'text': 'dangerous', 'id': 'DRRI-2 - A: Dangerous Military Duties'}}
+#8aa25d264897bd007d389890b2239c2b9c07fa0b: {'relation': 'drug used for treatment', 'head_span': {'text': 'High fever', 'id': 'High Grade Fever'}, 'tail_span': {'text': 'paracetamol', 'id': 'Acetaminophen Measurement'}}
+#d91bef9bfc94439523675b5d6a62e1f4635c0cdd: {'relation': 'medical condition treated', 'head_span': {'text': 'paracetamol', 'id': 'Acetaminophen Measurement'}, 'tail_span': {'text': 'High fever', 'id': 'High Grade Fever'}}
+```
+You can see, that on the coreference step the "it" pronoun in the second sentance was replaced by the unambiguous "High fever" entity. 
+After that the rebel model has extracted the trios of subject, relation and object and mapped them to the NCIT model. Note, that the mapping here is far from perfect. For example, the entity 'dangerous' was mapped to the 'DRRI-2 - A: Dangerous Military Duties' in NCIT. this is because in our mapping procedure for simplisity we have chosen the first result for the term in the NCIT database. To improve this, one would need to develop a more complex mapping algorithm.
+
+## Storing the results
+
+The final subject, relation, and object trios can be stored as either a labeled property graph or as an RDF graph. The guidelines to store the results as a neo4j labeled property graph are given [here](https://towardsdatascience.com/extract-knowledge-from-text-end-to-end-information-extraction-pipeline-with-spacy-and-neo4j-502b2b1e0754). Here we will give an approach to store the results as an RDF graph by using rdflib library in python.
+
+Rdflib allows the creation of entities with known URIs with the URIRef command. Also, one can create a custom namespace with new entities and relations. 

```

Comment:
````suggestion
The final subject, predicate, object triples can either be stored as a labeled property graph or as an RDF graph.

Here, we will give an approach to store the results as an RDF graph by using [rdflib]() library in python.

rdflib allows the creation of entities with known URIs with the URIRef command.
Also, one can create a custom namespace with new entities and relations. 

```{note}
For our audience interesting in labeled property graphs, the guidelines to store the results as a neo4j labeled property graph are available from  [this tutorial](https://towardsdatascience.com/extract-knowledge-from-text-end-to-end-information-extraction-pipeline-with-spacy-and-neo4j-502b2b1e0754).
```

````